Userland/Libraries/LibUnicode/Locale.cpp

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2021, Tim Flynn <trflynn89@pm.me>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/AllOf.h>
+#include <AK/CharacterTypes.h>
+#include <AK/QuickSort.h>
+#include <AK/StringBuilder.h>
+#include <LibUnicode/Locale.h>
+
+#if ENABLE_UNICODE_DATA
+#    include <LibUnicode/UnicodeLocale.h>
+#endif
+
+namespace Unicode {
+
+static String to_titlecase(const String& string)
+{
+    if (string.is_empty())
+        return string;
+
+    StringBuilder builder;
+    builder.append(to_ascii_uppercase(string[0]));
+
+    for (size_t i = 1; i < string.length(); ++i)
+        builder.append(to_ascii_lowercase(string[i]));
+
+    return builder.build();
+}
+
+bool is_unicode_language_subtag(StringView subtag)
+{
+    // unicode_language_subtag = alpha{2,3} | alpha{5,8};
+    if ((subtag.length() < 2) || (subtag.length() == 4) || (subtag.length() > 8))
+        return false;
+    return all_of(subtag, is_ascii_alpha);
+}
+
+bool is_unicode_script_subtag(StringView subtag)
+{
+    // unicode_script_subtag = alpha{4} ;
+    if (subtag.length() != 4)
+        return false;
+    return all_of(subtag, is_ascii_alpha);
+}
+
+bool is_unicode_region_subtag(StringView subtag)
+{
+    // unicode_region_subtag = (alpha{2} | digit{3}) ;
+    if (subtag.length() == 2)
+        return all_of(subtag, is_ascii_alpha);
+    if (subtag.length() == 3)
+        return all_of(subtag, is_ascii_digit);
+    return false;
+}
+
+bool is_unicode_variant_subtag(StringView subtag)
+{
+    // unicode_variant_subtag = (alphanum{5,8} | digit alphanum{3}) ;
+    if ((subtag.length() >= 5) && (subtag.length() <= 8))
+        return all_of(subtag, is_ascii_alphanumeric);
+    if (subtag.length() == 4)
+        return is_ascii_digit(subtag[0]) && all_of(subtag.substring_view(1), is_ascii_alphanumeric);
+    return false;
+}
+
+Optional<LanguageID> parse_unicode_language_id(StringView language)
+{
+    // https://unicode.org/reports/tr35/#Unicode_language_identifier
+    // unicode_language_id = "root"
+    //                       | (unicode_language_subtag (sep unicode_script_subtag)?
+    //                           | unicode_script_subtag)
+    //                         (sep unicode_region_subtag)?
+    //                         (sep unicode_variant_subtag)* ;
+    LanguageID language_id {};
+
+    if (language == "root"sv) {
+        language_id.is_root = true;
+        return language_id;
+    }
+
+    auto segments = language.split_view_if([](char ch) {
+        // sep = [-_] ;
+        return (ch == '-') || (ch == '_');
+    },
```

Comment:
I think we have `is_any_of("-_"sv)` that returns this lambda.

Reply:
Do we? We have `StringView::is_one_of` but that wouldn't work as the unary predicate here

Reply:
Here it is: 

https://github.com/SerenityOS/serenity/blob/cf271183b4befeba7c96bd6ca25246b899e836bb/AK/GenericLexer.h#L218

Reply:
Ah, nice!